[#396] Read and set LogLevel from environment variable

zmostafa opened the pull request:
## Notes for Reviewer
<!-- Items in addition to the checklist below that the reviewer should look for -->

Adds functionality to set the log level from the environment.

I went for @elBoberido proposal solution, where I read an environment variable called `IOX2_LOG_LEVEL` and set the LogLevel, with its value at runtime.

## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked in the [References](#references) section
1. [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
1. [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

<!-- Use either 'Closes #123' or 'Relates to #123' to reference the corresponding issue. -->

Closes #396 <!-- Add issue number after '#' -->

xieyuschen commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -179,6 +188,27 @@ pub enum LogLevel {
+impl FromStr for LogLevel {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```
I have wrote a part of it this afternoon, but you go ahead. Cool!

if it never returns an error, why not return LogLevel directly?

xieyuschen commented:
Moreover, please follow the convention here:

- title: `[#396]:  Read and set LogLevel from environment variable`

xieyuschen commented on iceoryx2-bb/log/src/logger/console.rs:
```diff
@@ -36,7 +36,9 @@ impl Default for Logger {
 impl Logger {
-    pub const fn new() -> Self {
+    pub fn new() -> Self {
+        init_log_level();
```
i think it's better to do lazy initialization in `get_log_level` and ensure all loggers get level by `get_log_level`. needn't call `init_log_level` everywhere.

zmostafa replied:
@xieyuschen Thanks, that's much better indeed.

elfenpiff commented:
@zmostafa hey hey ... you are back. I am happy to see your PR plopping up here.

zmostafa commented:
@elfenpiff you have to thank Jeff for the motivation, and thanks for the nice words as expected from you.

You guys are the best :heart:

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -179,6 +186,20 @@ pub enum LogLevel {
 }
 
+impl LogLevel {
+    fn from_str(s: &str) -> LogLevel {
+        match s {
+            "Trace" => LogLevel::Trace,
+            "Debug" => LogLevel::Debug,
+            "Info" => LogLevel::Info,
+            "Warn" => LogLevel::Warn,
+            "Error" => LogLevel::Error,
+            "Fatal" => LogLevel::Fatal,
+            _ => LogLevel::Info,
+        }
+    }
```
What do you think about making this a bit more relaxed by normalizing the input string to lower case. In this case the function name would be more like `from_str_fuzzy`.

I think it would be good to use `printf` and tell the user that the env variable contains an invalid string.

Finally, you can use `DEFAULT_LOG_LEVEL` for the default match arm.

zmostafa replied:
I had to change the return from the function to `u8` instead of `LogLevle` and updated the code accordingly

elBoberido replied:
Oh, you can also change the type of `DEFAULT_LOG_LEVEL` and `ENV_LOG_LEVEL` to `LogLevel`. Then there is only one `as u8` required :)

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -187,6 +208,7 @@ pub fn set_log_level(v: LogLevel) {
 
 /// Returns the current log level
 pub fn get_log_level() -> u8 {
+    LOG_LEVEL.store(*ENV_LOG_LEVEL as u8, Ordering::Relaxed);
```
Doesn't this prevent to set the log level programmatically?

I think you need to wrap this in a similar fashion like `INIT.call_once` and also do the call in `set_log_level`, similar to how it is done in `set_logger` but inside and outside of the `call_once` call.

zmostafa replied:
@elBoberido yes, calling it in `get_log_level` will override any value being set programmatically.
But do you meant calling it inside `get_logger` for once? I tried your proposal inside `get_log_level` but the init was not successful, but when i had it in `get_logger` just after `LOGGER` init, it seems to be the right place for init.

```
pub fn get_logger() -> &'static dyn Logger {
    INIT.call_once(|| {
        unsafe { LOGGER = Some(&*DEFAULT_LOGGER) };
        println!("setting env var");
        LOG_LEVEL.store(*ENV_LOG_LEVEL as u8, Ordering::Relaxed);
    });

    unsafe { *LOGGER.as_ref().unwrap() }
}

```

elBoberido replied:
Oh, I was not clear enough. You need to create a second static variable, e.g. `INIT_LOG_LEVEL`. The `INIT` variable could also be renamed to `INIT_LOGGER`.

So, no matter if `set_log_level` or `get_log_level` is called first, the `LOG_LEVEL` variable will be initialized only once with the entry from the env variable.

zmostafa replied:
@elBoberido Can you take a look on the last commit? I hope this is what you were asking for :)

zmostafa replied:
@elBoberido I moved the call to the functions mentioned above, now `INIT_LOG_LEVEL` is done in `set_log_level` and `get_log_level`

zmostafa replied:
@elBoberido I tested the changes on my terminal, even with setting an unknown trace level like `blabla` and printed a warning message to the user, and set the loglevel to default `Info` 
where should I document the new feature for debug LogLevel? FAQ file has a paragraph on setting log level programatically, should this get into the README.MD file?

elBoberido replied:
@zmostafa looks good. I think the FAQ would be a good candidate for this.

Could you please:
- also look into the other open discussion
- rebase to that we can trigger the CI
- try what happens when `IOX2_LOG_LEVEL` is set in the `env` section of `.cargo/config.toml` and also set as env variable in the terminal?
  - if the env variable set via the terminal has a higher priority than when set via `.cargo/config.toml`, then we could set the log level in `.cargo/config.toml` to trace, so that all iceoryx developer get the detailed log by default but can still control it by setting a custom level in `.bashrc`

zmostafa replied:
@elBoberido Branch is rebased.
setting IOX2_LOG_LEVEL in terminal takes precedence over setting it in `.cargo/config.toml` I added the env variable to `.cargon/config.toml` and configured it to Trace, also mentioned this precedence in FAQ.md

elBoberido replied:
@zmostafa it seems there is still a merge conflict in `iceoryx2-bb/log/src/lib.rs`. Can you do another rebase?

elBoberido replied:
@zmostafa okay, the branch is no mergeable. The markdown linter is now happy, though

elBoberido reviewed (commented):
@zmostafa great to have you back :)

I guess you need to sign the ECA. It can be done as individual contributor if you do the contributions in your spare time.

zmostafa commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -159,9 +161,14 @@ static DEFAULT_LOGGER: logger::tracing::Logger = logger::tracing::Logger::new();
+pub static ENV_LOG_LEVEL: Lazy<LogLevel> = Lazy::new(|| {
+    let log_level_str = env::var("IOX2_LOG_LEVEL").unwrap_or_else(|_| "Info".to_string());
+    LogLevel::from_str(&log_level_str) // Default to Info
+});
```
1. Can you explain more about string allocation? how do you track/bench them?
2. about the scenario you mentioned, is not it what should be in the first place? since we might run different instance of application from or within the same shell/env? so each application will utilize its logger?

elBoberido replied:
1. There would be [heaptrack](https://apps.kde.org/de/heaptrack/) as a nice gui to track all heap allocation. But in this case I just know that `String` will allocate on the heap from its documentation.
2. Each application will utilize its own logger instance but if the reading of the env variable is delayed, application A might read the value which was set after it was started and meant to be the value for application B.

elBoberido replied:
@zmostafa can you please revert the lazy initialization for the logger

zmostafa replied:
@elBoberido I reverted back the DEFAULT_LOGGER lazy initialization

elBoberido replied:
Can you do the same for `ENV_LOG_LEVEL`

orecham replied:
@zmostafa I think this alternative approach for reading the log level from the `env` encounters the same issue outlined above.

If the environment changes between application starting and `init_log_level()` being called, the application might use the wrong log level.

That being said, I do not immediately see a good way to prevent this and I think the previous approach with `Lazy` was actually better since the user didn't need to explicitly do anything and reading the `env` would likely occur earlier.

@elBoberido What did you have in mind as an alternative?

elBoberido replied:
@orecham the issue is the same but the time window where the env variable can be changed and influence a previously started process is minimal. With the lazy approach, this could be arbitrarily long

elBoberido reviewed (commented):
@zmostafa sorry for the slow response. I was quite busy the last days, but I try to improve on my response time :)

zmostafa commented on FAQ.md:
````diff
@@ -134,6 +134,20 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
 ## How to set the log level
 
+iceoryx2 gets the logging level from either the code, or the environment variable `IOX2_LOG_LEVEL`
+the developer can set this environment variable with one of the supported levels:
+```Trace, Debug, Info, Warning, Error, Fatal``` 
+
+`export IOX2_LOG_LEVEL=Trace`
+
+or via `.cargo/config.toml` 
````
@elBoberido no need to apologize, you are running a company now, how hard it can be? :D

elBoberido commented on FAQ.md:
```diff
@@ -58,24 +58,22 @@ insufficient, a new publisher with a larger `max_slice_len` can be created.
 
-> [!IMPORTANT]
-> Be aware that the history of the old publisher is lost when it is
+> [!IMPORTANT] Be aware that the history of the old publisher is lost when it is
```
Oh no, does the script have some autocorrection?
This breaks the github [alerts syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts). `> [!IMPORTANT]` must be on a separate line than the content

elBoberido replied:
@zmostafa can you please revert those changes. It breaks the alert syntax. Please also have a look at the other open point.

zmostafa replied:
@elBoberido hmmm, I think i did, are you sure you are looking at the latest commit?

elBoberido replied:
@zmostafa I hope to look at the latest commit :sweat_smile:

I did a Ctrl+F5, which should reload everything and dismiss caches but I still see `> [!IMPORTANT]` on the same line as `Be aware that ...`

orecham replied:
Note I believe this happens because the auto-corrector does not recognise GitHub-style admonitions, so you need to manually ensure these are on their own line otherwise the auto corrector will try and do line length correction with it on the same line.

Not ideal, I know, but the idea is for the corrector to fix around 90% of the issues which should save more time than manually fixing these small edge cases (e.g. manually fixing all line lengths).

elBoberido replied:
Can we skip line breaks in the auto-corrector? It's just a styling issue, not a rendering issue and it seems to break the github admonitions.

orecham replied:
I believe this is the line length corrector causing this. If disabled, then users would need to ensure line length limits were maintained, either manually or with an IDE. I deduced that this would be a larger annoyance for contributor than fixing these more limited edge cases. 

Of course, we could check if the auto corrector could be updated to recognise the GitHub admonitions (these were not used in our documentation at the time of preparing the script).

elBoberido replied:
Well, I would argue that one is not only an annoyance but breaks the rendering of the readme if the reviewer does not pay attention and the other one is a choice by ourselves and does not break the rendering if the line is too long. I would rather deactivate the line length detection than having to keep in mind to check for broken admonitions in the PR.

orecham replied:
Will fix the linter after this PR: https://github.com/eclipse-iceoryx/iceoryx2/issues/637

elBoberido commented:
@zmostafa I just wanted to check if you're waiting on anything from our side. If you need more time before continuing with the PR, feel free to take as much time as you need :)

elBoberido commented:
@zmostafa we would really like to have this PR merged. Are you still planing to finish the remaining issues? It's totally fine if you need more time or decided to move to other tasks but it would be great to get an update from you :)

zmostafa commented:
@elBoberido I am so sorry for ghosting this MR like this, I am not happy with my attitude here, please accept my apology :) 
I tried to refactor the merge request based on my memory of what is required, since I removed the lazy init, i added a new method `init_log_level` which will read the env var at application start and set the log level, i tested it locally but did not commit the example code for that.
Also. set_log_level take precedence when called after it to set the log level, hope my understanding is correct here, if not, happy to correct it. 


@orecham Thanks for getting in touch, i owe you much man!

orecham commented on FAQ.md:
````diff
@@ -105,17 +105,40 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
+
+`export IOX2_LOG_LEVEL=Trace`
+
+then in the main.rs call the function to read the environment variable
+```rust
````
```suggestion
then in the main.rs call the function to read the environment variable

```rust
```

zmostafa replied:
Done

orecham reviewed (commented):
The FreeBSD CI issue is not related to this PR - you can ignore it. Maybe also the Windows issue. 

Once the remaining open conversation is resolved I will help take care of the CI issues.

zmostafa commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +188,34 @@ pub enum LogLevel {
+            "error" => LogLevel::Error,
+            "fatal" => LogLevel::Fatal,
+            _ => {
+                println!("Error: you are using unknown logging level {:?}", s);
```
@orecham Done

zmostafa commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +188,34 @@ pub enum LogLevel {
+    let log_level = env::var("IOX2_LOG_LEVEL")
+        .ok()
+        .map(|s| LogLevel::from_str_fuzzy(&s))
+        .unwrap_or(LogLevel::Info);
```
Done

orecham commented:
> @orecham Regarding your last comment, I do not know why you see the old implementation for `ENV_LOG_LEVEL` it should have been replaced with `pub static ENV_LOG_LEVEL: OnceLock<LogLevel> = OnceLock::new();` in the new implementation

@zmostafa 

The last comment is actually part of this thread: https://github.com/eclipse-iceoryx/iceoryx2/pull/474#discussion_r1983165595

I did see your new version - it seems to have the same problem as the old one though, or am I misunderstanding?

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +188,34 @@ pub enum LogLevel {
+}
+
+/// Sets the log level by reading environment variable "IOX2_LOG_LEVEL"
+pub fn init_log_level() {
```
What do you think about renaming this to `init_log_level_from_env_or_default` to make it more clear what happens. Since there is already a `set_log_level` function, it might even make sense to call this `set_log_level_from_env_or_default` to make it more consistent. Finally, there could be a `set_log_level_from_env_or(log_level)` function which combines the two and set's the log_level only when the env variable is not set.

zmostafa replied:
@elBoberido Done and rebsed my branch to main to squash the commits for a cleaner histroy

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +188,34 @@ pub enum LogLevel {
+            "fatal" => LogLevel::Fatal,
+            _ => {
+                println!("Error: unknown logging level {:?}", s);
+                println!("Warning: setting log level as : Info");
+                DEFAULT_LOG_LEVEL
```
```suggestion
                println!("Warning: setting log level as : {}", DEFAULT_LOG_LEVEL);
                DEFAULT_LOG_LEVEL
```
I hope this works :sweat_smile: 
If not, then you would need to implement `Display`

zmostafa replied:
@elBoberido using `{:?}` for printing `DEFAULT_LOG_LEVEL` worked

```
Error: unknown logging level "BlaBla"
Warning: setting log level as : Info
Trigger event with id 1 ...
```

zmostafa replied:
@elBoberido makes sense, I changed it to 
```
println!(
                    "Invalid value for 'IOX_LOG_LEVEL' environment variable!\
                \nFound: {:?}\
                \nAllowed is one of: fatal, error, warn, info, debug, trace\
                \nSetting log level as : {:?}",
                    s, DEFAULT_LOG_LEVEL
                );
```

elBoberido reviewed (commented):
Looks good. There is an open PR to fix the FreeBSD and Windows build issue. Once merged, you can rebase to make the CI happy.

elBoberido commented on FAQ.md:
```diff
@@ -105,17 +105,41 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
+## Supported log levels
+
+iceoryx2 support different log levels
+`Trace, Debug, Info, Warning, Error, Fatal`
```
Just noticed, in iceoryx1 we used all lower space strings. For consistency, we probably should do the same in iceoryx2. Of course, we can keep the behavior that variants like `TrAcE` also work.

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +188,45 @@ pub enum LogLevel {
+
+/// Sets the log level by reading environment variable "IOX2_LOG_LEVEL" or default it wiht LogLevel::INFO
+pub fn set_log_level_from_env_or_default() {
+    let log_level = env::var("IOX2_LOG_LEVEL")
+        .ok()
+        .map(|s| LogLevel::from_str_fuzzy(&s))
+        .unwrap_or(DEFAULT_LOG_LEVEL);
+    ENV_LOG_LEVEL.set(log_level).unwrap();
+    set_log_level(*ENV_LOG_LEVEL.get().unwrap());
```
Please just delegate to `set_log_level_from_env_or`

zmostafa replied:
@elBoberido less code :sweat_smile:

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +188,45 @@ pub enum LogLevel {
+        .ok()
+        .map(|s| LogLevel::from_str_fuzzy(&s))
+        .unwrap_or(v);
+    ENV_LOG_LEVEL.set(log_level).unwrap();
+    set_log_level(*ENV_LOG_LEVEL.get().unwrap());
```
With the latest changes, I think you can remove `ENV_LOG_LEVEL` entirely and just use
```suggestion
    set_log_level(log_level);
```

elBoberido commented on FAQ.md:
````diff
@@ -105,17 +105,41 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
+
+then in the main.rs call the function to read the environment variable
+
+```rust
+use iceoryx2::prelude::*
+
+// ...
+
+set_log_level_from_env_or_default();
+```
+
+Developers can also set the logging level from within the code which will take
+precedence over the environment variable:
+
 ```rust
 use iceoryx2::prelude::*
````
What do you think of adding all three options to set the log level in the same code block? Maybe by adding an additional comment above each option?

elBoberido commented on iceoryx2-bb/log/Cargo.toml:
```diff
@@ -21,3 +21,4 @@ iceoryx2-pal-concurrency-sync = { workspace = true }
 termsize = { workspace = true }
 log = { workspace = true, optional = true }
 tracing = { workspace = true, optional = true }
+once_cell = { workspace = true }
```
I think this can be removed

elBoberido commented on iceoryx2-bb/log/BUILD.bazel:
```diff
@@ -37,5 +37,6 @@ rust_library(
     deps = [
         "//iceoryx2-pal/concurrency-sync:iceoryx2-pal-concurrency-sync",
         "@crate_index//:termsize",
+        "@crate_index//:once_cell",
```
This also

elBoberido reviewed (commented):
Thanks for your patience. It seems that with each of your commits, there are some new suggestions from our side, but this is just nitpicking at a high level :)

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +187,44 @@ pub enum LogLevel {
+            "fatal" => LogLevel::Fatal,
+            _ => {
+                println!(
+                    "Invalid value for 'IOX_LOG_LEVEL' environment variable!\
```
nitpick :)
```suggestion
                    "Invalid value for 'IOX2_LOG_LEVEL' environment variable!\
```

zmostafa replied:
:face_with_peeking_eye: copy&paste

elBoberido commented on iceoryx2-bb/log/src/lib.rs:
```diff
@@ -185,6 +187,44 @@ pub enum LogLevel {
+                \nFound: {:?}\
+                \nAllowed is one of: fatal, error, warn, info, debug, trace\
+                \nSetting log level as : {:?}",
+                    s, DEFAULT_LOG_LEVEL
+                );
+                DEFAULT_LOG_LEVEL
```
At this location we do not know which log level the user has chosen as fallback, the default one or a custom one. A workaround could be to pass the fallback log level to this function or just skip the actual value and print that the fallback log level is used.

elBoberido commented on FAQ.md:
```diff
@@ -105,31 +105,54 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
 
 ## How to set the log level
 
+While developing, iceoryx2 gets its default logging level from `.cargo/config.toml`, 
+but this can be over-ridden by using environment variable `IOX2_LOG_LEVEL` that can
+be set with one of the supported levels.
```
Sorry being picky about this. The default log level is defined in `iceoryx2-bb/log/src/lib.rs`. The `.cargo/config.toml` file is just a convenient way to set the environment variable for the specific project. If the environment variable was explicitly set by the user, that one will have precedence over the one in the toml file.

Since this section is not only about the environment variable, but in general about setting the log level, the emphasis should be about setting the log level programmatically and also with the environment variable. It can be mentioned that for convenience the environment variable can be set in the toml file, but that's only important for developers since that won't work in a deployment, which is probably what they are looking for when reading this document. Note, for iceoryx developers, there is the FAQ_ICEORYX2_DEVS.md.

zmostafa replied:
@elBoberido your repo, your picky rules
If I understand this correctly, I should remove the part that talks about setting it in the toml file, and focus only on the environment variable, right?

elBoberido commented on FAQ.md:
```diff
@@ -105,31 +105,54 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
+
+`export IOX2_LOG_LEVEL=Trace`
+
+then in the `main.rs` call the function to read the environment variable
```
```suggestion
then in the `main.rs` call one of the functions to set the log level
```

elBoberido commented on FAQ.md:
```diff
@@ -105,31 +105,54 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
 // ...
 
+// Reads LogLevel from env and defaults to LogLevel INFO
+// if the environemnt variable has an unsupported value
```
Small typo. Also, if the env variable is not set, the default log level will be used.
```suggestion
// if the environment variable is not set or has an unsupported value
```

elBoberido commented on FAQ.md:
```diff
@@ -105,31 +105,54 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
+set_log_level_from_env_or_default();
+
+// Reads LogLevel from env and sets it with a user-given value
+// if the environemnt variable has an unsupported value 
```
Same here :)
```suggestion
// if the environment variable is not set or has an unsupported value 
```

elBoberido commented on FAQ.md:
```diff
@@ -105,31 +105,54 @@ But you can also use a crate like [ctrlc](https://docs.rs/ctrlc/latest/ctrlc/).
+set_log_level_from_env_or(LogLevel::DEBUG);
+
+// sets LogLevel programmatically with a supported user-given value
+// this will take precedence over LogLevel defined by an environment variable
```
This might be ambiguous. One could assume that the env variable is still read but not used. What do you think about just `and does not try to read the environment variable`

elBoberido reviewed (commented):
Okay, I think the next round will be merged :)

elBoberido reviewed (dismissed):
Thanks for your patience. We could merge the PR but it seems there is a merge conflict in a file. Could you please rebase to the latest main?

zmostafa commented:
@elBoberido My patience? :D How should i thank you for yours then?

elBoberido reviewed (approved):
Okay, this will be the next one to be merged once the CI is finished

elBoberido commented:
@zmostafa it seems the CI is fighting to keep the PR open :D

The markdown linter is complaining

> FAQ.md:124:81 MD013/line-length Line length [Expected: 80; Actual: 149]
> FAQ.md:148:81 MD013/line-length Line length [Expected: 80; Actual: 235]

elBoberido commented:
@zmostafa congrats to your first PR on iceoryx2. The next ones wil be easier. If you like, you could expose the two new functions also to the C and C++ bindings and also adjust the examples
